File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /*
    If it's greater than the Max, then it's also greater than the Node
    before it. Using max is a good way of keeping track, that the
    current Node is greater than all that came before it.
    */
    int count = 0;
    int maxVal = 0;

    public int goodNodes(TreeNode root) {
        maxVal = root.val;
        dfs(root);
        return count;
    }

    private void dfs(TreeNode root) {
        if (root == null) {
            return;
        }
        
        count += (root.val > maxVal) ? 1 : 0; // current root greater than previous
        maxVal = Math.max(maxVal, root.val);
        count += dfs(root.left);
        count += dfs(root.right);
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:

    def __init__(self):
        self.balanced = True

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        temp = self.helper(root)
        return self.balanced
    
    def helper(self, node: Optional[TreeNode]) -> int:
        if node == None:
            return 0
        height = 0
        leftHeight = self.helper(node.left)
        rightHeight = self.helper(node.right)
        if abs(leftHeight - rightHeight) > 1:
            self.balanced = False
        return 1 + max(leftHeight, rightHeight)


"""
- check each and every node individually
- chain of thought is that each node should be treated like its own individual subtree with
all necessary properties like height, width, etc.
"""